docs/devcontainerjson-reference.md

```diff
@@ -0,0 +1,172 @@
+| `features` | object | An object of [dev container features](https://code.visualstudio.com/docs/remote/containers#_dev-container-features-preview) and related options to be added into your primary container. The specific options  that are available varies by feature, so see its documentation for additional details. For example: <br />`"features": {"github-cli": "latest"}` <br /><br /> ⚠️ Currently in preview. |
+| `shutdownAction` | enum | Indicates whether `devcontainer.json` supporting tools should stop the containers when the related tool window is closed / shut down.<br>Values are  `none`, `stopContainer` (default for image or Dockerfile), and `stopCompose` (default for Docker Compose).<br /><br /> ⚠️ Does not apply to Codespaces. |
+
+## VS Code specific properties
+
+While most properties apply to any `devcontainer.json` supporting tool or service, a few are specific to VS Code.
+
+| Property | Type | Description |
+|----------|------|-------------|
+| `extensions` | array | An array of extension IDs that specify the extensions that should be installed inside the container when it is created. Defaults to `[]`. |
+| `settings` | object | Adds default `settings.json` values into a container/machine specific settings file. Defaults to `{}`. |
+| `devPort` | integer | Allows you to force a specific port that VS Code Server should use in the container. Defaults to a random, available port. |
+
+## Lifecycle scripts
+
+When creating or working with a dev container, you may need different commands to be run at different points in the container's lifecycle. The table below lists a set of command properties you can use to update what the container's contents in the order in which they are run (for example, `onCreateCommand` will run after `initializeCommand`). Each command property is an string or list of command arguments that should execute from the `workspaceFolder`.
+
+| Property | Type | Description |
+|----------|------|-------------|
+| `initializeCommand` | string,<br>array | A command string or list of command arguments to run on the **host machine** before the container is created. .<br /><br /> ⚠️ The command is run wherever the source code is located on the host. For Codespaces, this is in the cloud. |
+| `onCreateCommand` | string,<br>array | This command is the first of three (along with `updateContentCommand` and `postCreateCommand`) that finalizes container setup when a dev container is created. It and subsequent commands execute **inside** the container immediately after it has started for the first time.<br /><br> GitHub Codespaces also use this command when generating prebuilt codespaces, but it executes before the container has been assigned to a specific user. It therefore can only take advantage of repository and org scoped secrets or permissions.|
```

Comment:
These next few lines reference GitHub Codespaces directly. Perhaps we shift these to more passive statements. e.g.

"Cloud services can use this command when caching or prebuilding a container. This means that it will not typically have access to user-scoped assets or secrets."

Reply:
Makes sense, thanks for the suggestion! Updated the descriptions in the lifecycle section to follow this model. Let me know if you have any other thoughts.